File: bench/cross-language/json_parse/json_parse.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NUM_ENTRIES 10000
#define KEY_SIZE 16
#define VALUE_SIZE 16

typedef struct {
    char key[KEY_SIZE];
    char value[VALUE_SIZE];
} Entry;

Entry* build_entries(int count) {
    Entry* entries = (Entry*)malloc(count * sizeof(Entry));
    for (int i = 0; i < count; i++) {
        snprintf(entries[i].key, KEY_SIZE, "key_%d", i);
        snprintf(entries[i].value, VALUE_SIZE, "value_%d", i);
    }
    return entries;
}

const char* find_entry(Entry* entries, int count, const char* target) {
    for (int i = 0; i < count; i++) {
        if (strcmp(entries[i].key, target) == 0) {
            return entries[i].value;
        }
    }
    return "";
}

int main() {
    Entry* entries = build_entries(NUM_ENTRIES);
    const char* found = find_entry(entries, NUM_ENTRIES, "key_9999");
    printf("Found: %s\n", found);
    printf("Count: %d\n", NUM_ENTRIES);
    free(entries);
    return 0;
}

File: bench/cross-language/json_parse/json_parse.zig
const std = @import("std");

const Entry = struct {
    key: [16]u8 = undefined,
    value: [16]u8 = undefined,
    key_len: usize = 0,
    value_len: usize = 0,
};

fn writeStr(buf: []u8, prefix: []const u8, num: usize) usize {
    var pos: usize = 0;
    for (prefix) |c| {
        buf[pos] = c;
        pos += 1;
    }
    // Write number digits
    if (num == 0) {
        buf[pos] = '0';
        pos += 1;
        return pos;
    }
    var n = num;
    var digits: [10]u8 = undefined;
    var dlen: usize = 0;
    while (n > 0) {
        digits[dlen] = @intCast(n % 10 + '0');
        dlen += 1;
        n /= 10;
    }
    var i: usize = dlen;
    while (i > 0) {
        i -= 1;
        buf[pos] = digits[i];
        pos += 1;
    }
    return pos;
}

fn strEql(a: []const u8, b: []const u8) bool {
    if (a.len != b.len) return false;
    for (a, b) |ac, bc| {
        if (ac != bc) return false;
    }
    return true;
}

pub fn main() !void {
    const count = 10_000;
    var entries: [count]Entry = undefined;

    for (0..count) |i| {
        entries[i].key_len = writeStr(&entries[i].key, "key_", i);
        entries[i].value_len = writeStr(&entries[i].value, "value_", i);
    }

    // Find key_9999
    var found: []const u8 = "";
    const target = "key_9999";
    for (0..count) |i| {
        if (strEql(entries[i].key[0..entries[i].key_len], target)) {
            found = entries[i].value[0..entries[i].value_len];
            break;
        }
    }

    const stdout_file = std.fs.File.stdout();
    var buf: [4096]u8 = undefined;
    var w = stdout_file.writer(&buf);
    try w.interface.print("Found: {s}\n", .{found});
    try w.interface.print("Count: {d}\n", .{count});
    try w.interface.flush();
}
